### ISSUE TITLE
Error while using TypeScript 2.4.1

### ISSUE THREAD

**Issue by elderapo** [samchon/tstl#5]
There were no errors when I used this lib with TypeScript 2.3.4. Besides that this lib is awesome and I love it <3
```
node_modules/tstl/lib/tstl.d.ts(4508,20): error TS2415: Class 'SetContainer<T>' incorrectly extends base class 'Container<T>'.
  Types of property 'begin' are incompatible.
    Type '() => SetIterator<T>' is not assignable to type '() => Iterator<T>'.
      Type 'SetIterator<T>' is not assignable to type 'Iterator<T>'.
        Types of property 'source' are incompatible.
          Type '() => SetContainer<T>' is not assignable to type '() => Container<T>'.
            Type 'SetContainer<T>' is not assignable to type 'Container<T>'.
              Types of property 'erase' are incompatible.
                Type '{ (val: T): number; (it: SetIterator<T>): SetIterator<T>; (begin: SetIterator<T>, end: SetIterato...' is not assignable to type '{ (position: Iterator<T>): Iterator<T>; <U extends T>(begin: Iterator<U>, end: Iterator<U>): I
ter...'.
                  Types of parameters 'val' and 'begin' are incompatible.
                    Type 'Iterator<any>' is not assignable to type 'T'.
node_modules/tstl/lib/tstl.d.ts(4712,11): error TS2415: Class '_SetElementList<T>' incorrectly extends base class '_ListContainer<T, SetIterator<T>>'.
  Types of property 'rbegin' are incompatible.
    Type '() => SetReverseIterator<T>' is not assignable to type '() => IReverseIterator<T>'.
      Type 'SetReverseIterator<T>' is not assignable to type 'IReverseIterator<T>'.
        Types of property 'base_' are incompatible.
          Type 'SetIterator<T>' is not assignable to type 'Iterator<T>'.
            Types of property 'source' are incompatible.
              Type '() => SetContainer<T>' is not assignable to type '() => Container<T>'.
                Type 'SetContainer<T>' is not assignable to type 'Container<T>'.
node_modules/tstl/lib/tstl.d.ts(4712,56): error TS2344: Type 'SetIterator<T>' does not satisfy the constraint '_ListIteratorBase<T>'.
  Types of property 'prev' are incompatible.
    Type '() => SetIterator<T>' is not assignable to type '() => _ListIteratorBase<T>'.
      Type 'SetIterator<T>' is not assignable to type '_ListIteratorBase<T>'.
        Types of property 'source' are incompatible.
          Type '() => SetContainer<T>' is not assignable to type '() => Container<T>'.
            Type 'SetContainer<T>' is not assignable to type 'Container<T>'.
              Types of property 'erase' are incompatible.
                Type '{ (val: T): number; (it: SetIterator<T>): SetIterator<T>; (begin: SetIterator<T>, end: SetIterato...' is not assignable to type '{ (position: Iterator<T>): Iterator<T>; <U extends T>(begin: Iterator<U>, end: Iterator<U>): I
ter...'.
                  Types of parameters 'val' and 'begin' are incompatible.
                    Type 'Iterator<any>' is not assignable to type 'T'.
node_modules/tstl/lib/tstl.d.ts(5695,20): error TS2415: Class 'MapContainer<Key, T>' incorrectly extends base class 'Container<Pair<Key, T>>'.
  Types of property 'begin' are incompatible.
    Type '() => MapIterator<Key, T>' is not assignable to type '() => Iterator<Pair<Key, T>>'.
      Type 'MapIterator<Key, T>' is not assignable to type 'Iterator<Pair<Key, T>>'.
        Types of property 'source' are incompatible.
          Type '() => MapContainer<Key, T>' is not assignable to type '() => Container<Pair<Key, T>>'.
            Type 'MapContainer<Key, T>' is not assignable to type 'Container<Pair<Key, T>>'.
              Types of property 'erase' are incompatible.
                Type '{ (key: Key): number; (it: MapIterator<Key, T>): MapIterator<Key, T>; (begin: MapIterator<Key, T>...' is not assignable to type '{ (position: Iterator<Pair<Key, T>>): Iterator<Pair<Key, T>>; <U extends Pair<Key, T>>(begin:
Ite...'.
                  Types of parameters 'key' and 'begin' are incompatible.
                    Type 'Iterator<any>' is not assignable to type 'Key'.
node_modules/tstl/lib/tstl.d.ts(6088,11): error TS2415: Class '_MapElementList<Key, T>' incorrectly extends base class '_ListContainer<Pair<Key, T>, MapIterator<Key, T>>'.
  Types of property 'rbegin' are incompatible.
    Type '() => MapReverseIterator<Key, T>' is not assignable to type '() => IReverseIterator<Pair<Key, T>>'.
      Type 'MapReverseIterator<Key, T>' is not assignable to type 'IReverseIterator<Pair<Key, T>>'.
        Types of property 'base_' are incompatible.
          Type 'MapIterator<Key, T>' is not assignable to type 'Iterator<Pair<Key, T>>'.
            Types of property 'source' are incompatible.
              Type '() => MapContainer<Key, T>' is not assignable to type '() => Container<Pair<Key, T>>'.
                Type 'MapContainer<Key, T>' is not assignable to type 'Container<Pair<Key, T>>'.
node_modules/tstl/lib/tstl.d.ts(6088,72): error TS2344: Type 'MapIterator<Key, T>' does not satisfy the constraint '_ListIteratorBase<Pair<Key, T>>'.
  Types of property 'prev' are incompatible.
    Type '() => MapIterator<Key, T>' is not assignable to type '() => _ListIteratorBase<Pair<Key, T>>'.
      Type 'MapIterator<Key, T>' is not assignable to type '_ListIteratorBase<Pair<Key, T>>'.
        Types of property 'source' are incompatible.
          Type '() => MapContainer<Key, T>' is not assignable to type '() => Container<Pair<Key, T>>'.
            Type 'MapContainer<Key, T>' is not assignable to type 'Container<Pair<Key, T>>'.
              Types of property 'erase' are incompatible.
                Type '{ (key: Key): number; (it: MapIterator<Key, T>): MapIterator<Key, T>; (begin: MapIterator<Key, T>...' is not assignable to type '{ (position: Iterator<Pair<Key, T>>): Iterator<Pair<Key, T>>; <U extends Pair<Key, T>>(begin:
Ite...'.
                  Types of parameters 'key' and 'begin' are incompatible.
                    Type 'Iterator<any>' is not assignable to type 'Key'.
node_modules/tstl/lib/tstl.d.ts(10850,11): error TS2415: Class 'MapIterator<Key, T>' incorrectly extends base class '_ListIteratorBase<Pair<Key, T>>'.
  Types of property 'prev' are incompatible.
    Type '() => MapIterator<Key, T>' is not assignable to type '() => _ListIteratorBase<Pair<Key, T>>'.
      Type 'MapIterator<Key, T>' is not assignable to type '_ListIteratorBase<Pair<Key, T>>'.
        Types of property 'source' are incompatible.
          Type '() => MapContainer<Key, T>' is not assignable to type '() => Container<Pair<Key, T>>'.
            Type 'MapContainer<Key, T>' is not assignable to type 'Container<Pair<Key, T>>'.
              Types of property 'erase' are incompatible.
                Type '{ (key: Key): number; (it: MapIterator<Key, T>): MapIterator<Key, T>; (begin: MapIterator<Key, T>...' is not assignable to type '{ (position: Iterator<Pair<Key, T>>): Iterator<Pair<Key, T>>; <U extends Pair<Key, T>>(begin:
Ite...'.
                  Types of parameters 'key' and 'begin' are incompatible.
                    Type 'Iterator<any>' is not assignable to type 'Key'.
node_modules/tstl/lib/tstl.d.ts(10915,81): error TS2344: Type 'MapContainer<Key, T>' does not satisfy the constraint 'Container<Pair<Key, T>>'.
  Types of property 'begin' are incompatible.
    Type '() => MapIterator<Key, T>' is not assignable to type '() => Iterator<Pair<Key, T>>'.
      Type 'MapIterator<Key, T>' is not assignable to type 'Iterator<Pair<Key, T>>'.
        Types of property 'source' are incompatible.
          Type '() => MapContainer<Key, T>' is not assignable to type '() => Container<Pair<Key, T>>'.
            Type 'MapContainer<Key, T>' is not assignable to type 'Container<Pair<Key, T>>'.
              Types of property 'erase' are incompatible.
                Type '{ (key: Key): number; (it: MapIterator<Key, T>): MapIterator<Key, T>; (begin: MapIterator<Key, T>...' is not assignable to type '{ (position: Iterator<Pair<Key, T>>): Iterator<Pair<Key, T>>; <U extends Pair<Key, T>>(begin:
Ite...'.
                  Types of parameters 'key' and 'begin' are incompatible.
                    Type 'Iterator<any>' is not assignable to type 'Key'.
node_modules/tstl/lib/tstl.d.ts(10948,11): error TS2415: Class 'SetIterator<T>' incorrectly extends base class '_ListIteratorBase<T>'.
  Types of property 'prev' are incompatible.
    Type '() => SetIterator<T>' is not assignable to type '() => _ListIteratorBase<T>'.
      Type 'SetIterator<T>' is not assignable to type '_ListIteratorBase<T>'.
        Types of property 'source' are incompatible.
          Type '() => SetContainer<T>' is not assignable to type '() => Container<T>'.
            Type 'SetContainer<T>' is not assignable to type 'Container<T>'.
              Types of property 'erase' are incompatible.
                Type '{ (val: T): number; (it: SetIterator<T>): SetIterator<T>; (begin: SetIterator<T>, end: SetIterato...' is not assignable to type '{ (position: Iterator<T>): Iterator<T>; <U extends T>(begin: Iterator<U>, end: Iterator<U>): I
ter...'.
                  Types of parameters 'val' and 'begin' are incompatible.
                    Type 'Iterator<any>' is not assignable to type 'T'.
node_modules/tstl/lib/tstl.d.ts(11006,65): error TS2344: Type 'SetContainer<T>' does not satisfy the constraint 'Container<T>'.
  Types of property 'begin' are incompatible.
    Type '() => SetIterator<T>' is not assignable to type '() => Iterator<T>'.
      Type 'SetIterator<T>' is not assignable to type 'Iterator<T>'.
        Types of property 'source' are incompatible.
          Type '() => SetContainer<T>' is not assignable to type '() => Container<T>'.
            Type 'SetContainer<T>' is not assignable to type 'Container<T>'.
              Types of property 'erase' are incompatible.
                Type '{ (val: T): number; (it: SetIterator<T>): SetIterator<T>; (begin: SetIterator<T>, end: SetIterato...' is not assignable to type '{ (position: Iterator<T>): Iterator<T>; <U extends T>(begin: Iterator<U>, end: Iterator<U>): I
ter...'.
                  Types of parameters 'val' and 'begin' are incompatible.
                    Type 'Iterator<any>' is not assignable to type 'T'.
```

**Comment by samchon** [comment]
Thanks for your reporting. As you mentioned, such errors haven't occurred until the TypeScript 2.4 update. I need to check it out the origin of those errors.

I will write an issue in the repository of TypeScript arguing those errors. If it's a domain error of TypeScript 2.4, then I will wait the next update. If it's not an error, but a type of break change that TypeScript desired, then I will fix my library.

**Comment by samchon** [comment]
This error has fixed since the 1.5.3 update.
